viewing_party/party.py

```diff
@@ -1,23 +1,162 @@
+        if watchlist[item]["title"] == title:
+            user_data["watched"].append(watchlist[item])
+            del watchlist[item]
+
+    return user_data
 
 # -----------------------------------------
 # ------------- WAVE 2 --------------------
 # -----------------------------------------
+def get_watched_avg_rating(user_data):
+    sum = 0
+    num_of_ratings = 0
+    watched_list = user_data["watched"]
+    for dictionary in watched_list:
+        sum += dictionary["rating"]
+        num_of_ratings += 1
+    if num_of_ratings == 0:
+        return 0
+    else:
+        rating_avg = sum / num_of_ratings
 
+    return rating_avg
+
+def get_most_watched_genre(user_data):
+    if len(user_data["watched"]) < 1:
+        return None
+    genres = []
+    for dictionary in user_data["watched"]:
+        genres.append(dictionary["genre"])
+    genre_set = set(genres)
+    top_genre_dict = {genre : genres.count(genre) for genre in genre_set}
+    counter = 0
+    for value in top_genre_dict.values():
+        if counter < value:
+            counter = value
+            top_genre = list(top_genre_dict.keys())[list(top_genre_dict.values()).index(counter)]
+        
+    return top_genre
 
 # -----------------------------------------
 # ------------- WAVE 3 --------------------
 # -----------------------------------------
+def get_unique_watched(user_data):
+    user_movie_list = user_data["watched"]
+    friend_watched_list = user_data["friends"]
+    user_unique_movies = []
+    #------------------user data-------------------
+    for dictionary in user_movie_list:
+        if dictionary["title"] not in user_unique_movies:
+            user_unique_movies.append(dictionary)
+    
+    #--------------friend watched items-------------
+    for friends_dict in friend_watched_list:
+        for watched_list in friends_dict:
+            for dict in friends_dict[watched_list]:
+                for dictionary in user_unique_movies:
+                    if dictionary["title"] == dict["title"]:
+                        user_unique_movies.remove(dictionary)
+
+    return user_unique_movies
+
+def get_friends_unique_watched(user_data):
+    user_movie_list = user_data["watched"]
+    friend_watched_list = user_data["friends"]
+    friends_unique_movies = []
+
+    for friends_dict in friend_watched_list:
+        for dict in friends_dict["watched"]:
+            if dict["title"] == None:
+                return friends_unique_movies
+            if dict not in friends_unique_movies:
+                #### end up getting 4 instead of 3 movies in the final dictionary###
+                friends_unique_movies.append(dict)
+
+    for dictionary in user_movie_list:
+        for dict in friends_unique_movies:
+            if dictionary["title"] == dict["title"]:
+                #########does not remove common titles between user & friend. Look into it######
+                friends_unique_movies.remove(dict)
+
+    return friends_unique_movies

```

Comment:
For this function you can do the opposite to what I proposed in the function prior. You can create a list to add or remove from based on whether or not the user and their friends have watched it.